### PR TITLE
Move knapsack algorithms into separate filing (#141451)

### PR DESCRIPTION
**Pull request by basilwong**
Summary:

Original Context Doc: https://docs.google.com/document/d/1Gv5ZqN3UY7kTuCUd0JLU3L_onwVIr2vd3AiZyim1Muc/edit?disco=AAABZX_tqdk

### Changes

This diff restructures the Partitioners.py file in the _functorch package.

* Moves the three knapsack problem algorithems (greedy, ilp, dp) into a separate file

Test Plan:
### Unit Testing

```
$ buck test mode/opt //caffe2/test/functorch:test_ac
File changed: fbsource//xplat/caffe2/test/functorch/TARGETS
File changed: fbsource//xplat/caffe2/test/functorch
File changed: fbsource//xplat/caffe2/test
7 additional file change events
Soft Error: source_directory_includes_subpackage: Directory `v2.17.1-1` of package `fbsource//third-party/nccl` may not cover any subpackages, but includes subpackage `v2.17.1-1/src/tests`.
Soft Error: source_directory_includes_subpackage: Directory `v2.18.3-1` of package `fbsource//third-party/nccl` may not cover any subpackages, but includes subpackage `v2.18.3-1/src/tests`.
Soft Error: source_directory_includes_subpackage: Directory `v2.19.3-1` of package `fbsource//third-party/nccl` may not cover any subpackages, but includes subpackage `v2.19.3-1/src/tests`.
Buck UI: https://www.internalfb.com/buck2/a2f91f8a-5326-435e-9075-5af0de930b8b
Test UI: https://www.internalfb.com/intern/testinfra/testrun/7036874660108924
Network: Up: 28MiB  Down: 3.5GiB  (reSessionID-19af3d71-7528-448c-9126-5615d27b3bd7)
Jobs completed: 423656. Time elapsed: 3:57.2s.
Cache hits: 99%. Commands: 146147 (cached: 145758, remote: 317, local: 72)
Tests finished: Pass 8. Fail 0. Fatal 0. Skip 0. Build failure 0

```

### Tested on Local Training Run

```
CUDA_VISIBLE_DEVICES=5,6 AOT_PARTITIONER_DEBUG=1 PARTITIONER_MEMORY_BUDGET_PARETO=0 buck2 run mode/opt //aps_models/ads/icvr:icvr_launcher -- mode=local_fb_fm_v4 launcher.num_workers=2 2>&1 | tee log_2024-11-2320:41:50.757256__bento_trigger.txt
```

Output Summary Paste: P1685697066

Differential Revision: D65800097

### CODE REVIEW

**Comment by basilwong** [comment]
Previous PR: https://github.com/pytorch/pytorch/pull/141451

File: torch/_functorch/_activation_checkpointing/knapsack.py
Comment (jansel) [inline torch/_functorch/_activation_checkpointing/knapsack.py:1]: A single underscore in the path is sufficient to indicate it is private, so `_activation_checkpointing` can be `activation_checkpointing`.

File: torch/_functorch/_activation_checkpointing/knapsack.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
+# mypy: ignore-errors
```
Comment (jansel) [inline torch/_functorch/_activation_checkpointing/knapsack.py:1]: Why are we ignoring all mypy errors here?

**Review by jansel — CHANGES_REQUESTED** [review]
@Chillee should chime in on the best directory structure here.

File: torch/_functorch/_activation_checkpointing/knapsack.py (hunk 1 above)
Comment (basilwong) [reply torch/_functorch/_activation_checkpointing/knapsack.py:1]: oh, just copying the what other files in the directory where doing...will remove

File: torch/_functorch/_activation_checkpointing/knapsack.py
Comment (basilwong) [reply torch/_functorch/_activation_checkpointing/knapsack.py:1]: image

The reason I added the underscore is that I saw `_aot_autograd` in the same directory.

**Review by Chillee — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**